src/mlmc/quantity_concept.py

```diff
@@ -121,120 +325,451 @@ def mult_op(x, y):
 
+    def storage_id(self):
+        """
+        Identity of QuantityStorage instance
+        :return: int
+        """
+        return id(self)
 
+    def samples(self, level_id, i_chunk):
+        """
+        Get results for given level id and chunk id
+        :param level_id: int
+        :param i_chunk: int
+        :return: Array[M, chunk size, 2]
+        """
+        level_chunk = self._storage.sample_pairs_level(level_id, i_chunk)  # Array[M, chunk size, 2]
+        if level_chunk is not None:
+            assert self.qtype.size() == level_chunk.shape[0]
+            # Select values from given interval self.start:self.end
+            if self.start is not None and self.end is not None:
+                return level_chunk[self.start:self.end, :, :]
+        return level_chunk
+
+    def __copy__(self):
+        new = type(self)(self._storage, self.qtype)
+        new.__dict__.update(self.__dict__)
+        return new
+
+
+class QType(metaclass=abc.ABCMeta):
+    def size(self) -> int:
+        """
+        Size of type
+        :return: int
+        """
 
-class Dict(QBase):
-    def __init__(self, *args: Tuple[str, QBase]):
-        self.quantity_dict = OrderedDict(*args)
+    def base_qtype(self):
+        return self._qtype.base_qtype()
 
-    def __getattr__(self, key):
-        # JB: TODO
-        if key in self.quantity_dict:
-            return self.quantity_dict[key]
-        return getattr(key)
+    def __eq__(self, other):
+        if isinstance(other, QType):
+            return self.size() == other.size()
+        return False
 
-    def max(self):
+    def replace_scalar(self, new_qtype):
         """
-        Create quantity for the maximum over the time series.
-        :return:
+        Find ScalarType and replace it with new_qtype
+        :param new_qtype: QType
+        :return: None
         """
-        return _QMax(*[q for q in self.quantity_dict.values()])
+        if isinstance(self._qtype, ScalarType):
+            self._qtype = new_qtype
+        else:
+            self._qtype.replace_scalar(new_qtype)
+
+
+class ScalarType(QType):
+    def __init__(self, qtype=float):
+        self._qtype = qtype
+
+    def base_qtype(self):
+        return self
+
+    def size(self) -> int:
+        return 1
+
+
+class BoolType(ScalarType):
+    pass
+
+
+class ArrayType(QType):
+    def __init__(self, shape, qtype: QType, start=0):
+        self._shape = shape
+        self._qtype = qtype
+        self.start = start
+
+    def size(self) -> int:
+        return np.prod(self._shape) * self._qtype.size()
+
+    def __getitem__(self, key):
+        """
+        ArrayType indexing
+        :param key: int, tuple of ints or slice objects
+        :return: QuantityType - ArrayType or self._qtype
+        """
+        # Get new shape
+        new_shape = np.empty(self._shape)[key].shape
+
+        # One selected item is considered to be a scalar QType
+        if len(new_shape) == 1 and new_shape[0] == 1:
```

Comment:
We should keep the produced new_shape. Be as close to the numpy behavior as possible.